### DISCUSSION TITLE
Rev 23 'Ready Brief'

### DISCUSSION THREAD

**Discussion by SirNibble** [prirun/p50em#28] [General]
Not had much time in the  last few weeks, but have been flipping through PDFs to kick the ol' memory.

Going through them, discovered R23 really expanded on the functionality of Ready Brief'.

Now, why won't the following show the '>' (or '!' for error) character? 

`rdy -rb '%[translate %ui@%sn:%en '%lcase%' '%UCASE%' %]>' -expand_prompt `

I get the pathname change as expected, but just a space afterwards.

```
sman@pip01:sman pwd
<USERS>SMAN
sman@pip01:sman a *>plp
sman@pip01:plp ld

<USERS>SMAN>PLP (ALL access)
16 records in this directory, 16 total records out of quota of 0.

5 Files.

CP.BIN           CP.PLP           CP.RUN           LO4000
TEST.COMI

1 Segment Directory.

CP.SEG

sman@pip01:plp lad
Not found.  LAD (std$cp)
sman@pip01:plp                                    

```
What have I missed?

**Comment by drboone** [comment]
Pretty sure this is at least one, and maybe two, bugs in Rev23.3.  First, < and > are treated as magic somehow, and it's not possible to make that go away with quoting or ~ (suppress treewalking &c).  Second, anything after the %] is eaten.

Rev24 doesn't have the undocumented behavior around < and >, and the desired result is achieved there:
```
OK, rdy -rb '%[translate %ui@%sn:%en> abcdefghijklmnopqrstuvwxyz ABCDEFGHIJKLMNOPQRSTUVWXYZ%] %dt' -expand_prompt
guest@em24:guest24>
```
However, 24 still eats anything after the %], which is at best counterintuitive.  Getting a space after the > is left as an exercise for the reader. :)

**Comment by drboone** [comment]
Oh, and note that there is some interaction between RDY and ECL, which might cause additional unexpectation.  It's documented in the commands manual RDY section.

**Comment by hashbackup** [comment]
I tried it on R23 and it seemed to work fine.  I was just typing rdy commands without ecl.
```
OK, rdy -rb 'jim>'
jim> rdy -rb '%[translate abc def aaa]>'
%[translate abc def aaa]> rdy -rb '%[translate abc def aaa]>' -expand_prompt
$UNBALANCED_BRACKETS$rdy -rb '%[translate abc def aaa%]>' -expand_prompt
dbc>rdy -rb '%[translate abc def aaa %]>' -expand_prompt
dbc>rdy -rb '%[translate 'abc' 'def' 'aaa' %]>' -expand_prompt
dbc>
```
It is a bit curious why the first rdy command adds a space to the prompt, while the ones with translate and > don't add a space.  I thought maybe it was because of -expand_prompt, but it isn't:
```
OK, rdy -rb 'jim>'
jim> a
jim> rdy -rb 'jim>' -expand_prompt
jim> a
jim> 
```
Maybe it only adds the space if there is nothing fancy in the prompt.

**Comment by hashbackup** [comment]
Also, understanding quoting and unquoting with Prime command lines is just as confusing as it is on Unix.  Normally the [quote and [unquote functions are used to handle this.  You can't just add an extra set of quotes whenever you think it might be necessary.  For example:
```
jim> type 'abc 'def' ghi'
abc def ghi
jim> type [quote 'abc 'def' ghi']
'abc 'def' ghi'
jim> type ''abc 'def' ghi''
abc def ghi
jim> type [length x]
1
jim> type [length x y z]
5
jim> type [length 'x y z']
5
jim> type [length ''x y z'']
5
jim> type [length [quote x y z]]
5
jim> type [length [quote 'x y z']]
7
jim> 
```
Adding a set of extra quotes does not automatically quote everything.
NB: all of the quotes above are single quotes.

**Comment by drboone** [comment]
It seems to take both RDY prompt variables _and_ a %[ %] block to trigger it.

**Comment by drboone** [comment]
Also, `$UNBALANCED_BRACKETS$` seems to mean a lot of things besides those that obviously involve brackets.  Get a mismatched set of characters into the before/after sets for `TRANSLATE`, $UNBALANCED_BRACKETS$.  Use "asdf" (with quotes) as the prompt string inside translate, $UNBALANCED_BRACKETS$.  Etc.

**Comment by SirNibble** [comment]
Thanks!
